gcc darwin fixes

matthewbauer opened the pull request:
###### Motivation for this change

These changes seem to get gcc48, gcc5, and gcc6 working on Darwin. I need to verify that Linux gcc's still work though.
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

matthewbauer committed:
gcc: --enable-bootstrap on Darwin

This seems to be working now. ISL needs to be disable for it to build
completely though.

matthewbauer committed:
gcc: fix gmp, mpfr includes

This should get gcc48, gcc5, and gcc6 working again.

Also: use makeLibraryPath, and makeSearchPathOutput for LIBRARY_PATH and
CPATH. This is a refactor but it also fixes an issue with zlib.

dezgeg commented on pkgs/development/compilers/gcc/4.8/default.nix:
```diff
@@ -470,7 +472,7 @@ stdenv.mkDerivation ({
   LIBRARY_PATH = concatStrings
                    (intersperse ":" (map (x: x + "/lib")
-                                         (optionals (zlib != null) [ zlib ]
+                                         (optionals (zlib != null) [ zlib.out ]
```
`concatStrings (intersperse ":" (map (x: x + "/lib")` should be able to be shortened to `lib.makeLibraryPath`, then `.out` from `zlib.out` should be able to be dropped.

(I actually wonder how this even works on current `master` since these resolve to e.g. `${gmp.dev}/lib`).

globin commented:
Still getting:
```
The directory that should contain system headers does not exist:
  /usr/include
make[3]: *** [Makefile:2829: stmp-fixinc] Error 1
```

LnL7 commented:
The versions you mentioned still build on linux.

matthewbauer committed:
gcc: add --with-native-system-header-dir for clang

gcc needs to be able find system headers. Without this, gcc fails to build because
/usr/include is not available.

Note: stdenv.libc should be available for all stdenv's, I think.

dezgeg commented on pkgs/development/compilers/gcc/4.8/default.nix:
```diff
@@ -359,6 +359,8 @@ stdenv.mkDerivation ({
     }
     ${if (stdenv ? glibc && cross == null)
       then " --with-native-system-header-dir=${stdenv.glibc.dev}/include"
+      else if (stdenv ? libc && cross == null)
```
You should be able to change the existing check to
````
     ${if cross == null
￼        then " --with-native-system-header-dir=${lib.getDev stdenv.libc}/include"
````

which should do the right thing for all platforms.

matthewbauer commented:
Ok so I've got my Macbook running without the Command Line Tools. I'll see what is fixable.

matthewbauer committed:
gcc4.5: allow empty libc-ldflags

matthewbauer committed:
gcc4.6: remove from tree

This isn't available from all-packages, so I think it's okay to
remove. It's unclear how to keep this updated with the changes.

matthewbauer committed:
gcc: use special native system headers for darwin

Darwin systems need to be able to find CoreFoundation headers as well as
libc headers. Somehow, gcc doesn't accept any "framework" parameters
that would normally be used to include CoreFoundation in this
situation.

HACK: Instead, this adds a derivation that combines the two. The result
works but probably not a good long term solution.

ALTERNATIVES: Maybe sending patches in to GCC to allow
"native-system-framework" configure flag to get this found.

matthewbauer commented:
Does anyone know if gcc 4.6 can be deleted? It's not reachable right now in the code AFAICT.

dezgeg commented:
Most likely yes, probably cbb055c1d9e16e6333a199b6df0380c14af55825 forgot to remove it by mistake.

vcunat commented:
`stdenv.libc` doesn't exist on Linux. Still, `stdenv.cc.libc` should be portable AFAIK.

vcunat reviewed (approved):
Except for the evaluation problem it looks OK and default version builds fine on x86_64-linux.

vcunat commented:
(Note: in any case, it shouldn't be merged to master directly.)

vcunat commented:
Staged.
